content/en/blog/2023-01-30-building-a-raft-part-3.md

```diff
@@ -0,0 +1,110 @@
+
+Instead, a leader periodically (say, every `n` seconds or so) will send an empty `AppendEntry` request to all of the followers (e.g. one without any new log entries). Followers can then tell this is a heartbeat from the leader and reset their election timeout, preventing a new election and therefore a new term. Speaking of which, it’s time to move on to the final RPC in Raft.
+
+## Leader Election
+
+Leader election in Raft is handled by the `RequestVote` RPC. First off, remember at the start of this series we mentioned that there are three states that a Raft server can be in: Leader, Candidate and Follower. We haven’t talked about the candidate state yet because this is where it actually comes into play.
+
+!["State change flowchart"](/img/blog/2023-01-19-building-a-raft-part-3/stateChanges.png)
+
+If a Raft server does not receive any communication from the leader (heartbeat or regular requests) for a given amount of time, it will assume the leader has died and will start a campaign to become the new leader, incrementing it’s term, voting for itself and sending `RequestVote` RPCs to the other servers in the cluster. If it receives a majority of the cluster’s votes, it declares itself leader and starts sending out heartbeats. Luckily for us, the `RequestVote` RPC ends up being much simpler than `AppendEntries`, but we should still spend some time stepping through it.
+
+## RequestVote Steps
+
+The `RequestVote` function is how candidates carry out elections in remote replicas. The function expects as input a `VoteRequest`, which contains the candidate's term, it's unique identifier, and some details of it's last log entry.
+
+*Note: Abbreviated for brevity*
+```proto
+message VoteRequest {
+    int32 term = 2;           // The candidate's term.
+    string candidateId = 3;   // ID of candidate requesting a vote.
+    int64 lastLogIndex = 4;   // The index of candidate’s last log entry
+    int32 lastLogTerm = 5;    // The term number of candidate’s last log entry
+}
+```
+
+In return, each voter will respond with a `VoteReply`, which contains their own unique identifier, their perspective of the term, and their vote.
+
+```proto
+message VoteReply {
+    string id = 1;            // ID of the Raft node sending the vote.
+    int32 term = 2;           // Current term number
+    bool voteGranted = 3;     // True if the voter has voted for the candidate
+}
+```
+
+#### 1. Vote "no" if the candidate is behind
+
+Just like in `AppendEntries`, we first need to check if the candidate has an up-to-date term. If a candidate requesting a vote has a term less than the recipient's term, that candidate is not up-to-date with the rest of the cluster, and should be denied the vote.
```

Comment:
@DanielSollis - can you please spell out how specifically this is being done in the code? It looks like we're relying on Go to set the default value of `voteGranted` to false on line 83, which then gets returned to the candidate on line 102, but it would be considerate to our reader to spell this out so they don't have to guess about what's going on!